apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/server/service/ServerAppOpenApiService.java

```diff
@@ -18,36 +18,80 @@
+import com.ctrip.framework.apollo.common.exception.BadRequestException;
 import com.ctrip.framework.apollo.common.utils.BeanUtils;
 import com.ctrip.framework.apollo.openapi.api.AppOpenApiService;
 import com.ctrip.framework.apollo.openapi.dto.OpenAppDTO;
 import com.ctrip.framework.apollo.openapi.dto.OpenEnvClusterDTO;
 import com.ctrip.framework.apollo.openapi.util.OpenApiBeanUtils;
 import com.ctrip.framework.apollo.portal.component.PortalSettings;
+import com.ctrip.framework.apollo.portal.entity.model.AppModel;
 import com.ctrip.framework.apollo.portal.environment.Env;
+import com.ctrip.framework.apollo.portal.listener.AppCreationEvent;
 import com.ctrip.framework.apollo.portal.service.AppService;
 import com.ctrip.framework.apollo.portal.service.ClusterService;
+import com.ctrip.framework.apollo.portal.service.RoleInitializationService;
+import com.ctrip.framework.apollo.portal.spi.UserInfoHolder;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.LinkedList;
 import java.util.List;
+import org.springframework.context.ApplicationEventPublisher;
 import org.springframework.stereotype.Service;
 
 /**
  * @author wxq
  */
 @Service
 public class ServerAppOpenApiService implements AppOpenApiService {
+
   private final PortalSettings portalSettings;
   private final ClusterService clusterService;
   private final AppService appService;
 
+  private final ApplicationEventPublisher publisher;
+
+  private final RoleInitializationService roleInitializationService;
+
+  private final UserInfoHolder userInfoHolder;
+
   public ServerAppOpenApiService(
       PortalSettings portalSettings,
       ClusterService clusterService,
-      AppService appService) {
+      AppService appService, ApplicationEventPublisher publisher,
+      RoleInitializationService roleInitializationService, UserInfoHolder userInfoHolder) {
     this.portalSettings = portalSettings;
     this.clusterService = clusterService;
     this.appService = appService;
+    this.publisher = publisher;
+    this.roleInitializationService = roleInitializationService;
+    this.userInfoHolder = userInfoHolder;
+  }
+
+  private App convert(OpenAppDTO openAppDTO) {
+    return App.builder()
+        .appId(openAppDTO.getAppId())
+        .name(openAppDTO.getName())
+        .ownerName(openAppDTO.getOwnerName())
+        .orgId(openAppDTO.getOrgId())
+        .orgName(openAppDTO.getOrgName())
+        .ownerEmail(openAppDTO.getOwnerEmail())
+        .build();
+  }
+
+  /**
+   * @see com.ctrip.framework.apollo.portal.controller.AppController#create(AppModel)
+   */
+  @Override
+  public void createApp(OpenAppDTO openAppDTO) {
+    final String appId = openAppDTO.getAppId();
+
+    List<OpenAppDTO> openAppDTOList = this.getAppsInfo(Collections.singletonList(appId));
+    if (null != openAppDTOList && !openAppDTOList.isEmpty()) {
+      throw new BadRequestException("AppId " + appId + " exists already, cannot create again");
+    }
```

Comment:
I think this check is redundant as the same check is done in `com.ctrip.framework.apollo.portal.service.AppService#createAppInLocal`?

Reply:
Yes, remove the check here already.